source/differentiation/ad/ad_helpers.cc

```diff
@@ -1641,6 +1641,248 @@ namespace Differentiation
+        jacobian.reinit({this->n_dependent_variables(),
+                         this->n_independent_variables()},
+                        true /*omit_default_initialization*/);
+
+      if (ADNumberTraits<ad_type>::is_taped == true)
+        {
+          Assert(this->active_tape_index() !=
+                   Numbers<ad_type>::invalid_tape_index,
+                 ExcMessage("Invalid tape index"));
+          Assert(this->is_recording() == false,
+                 ExcMessage(
+                   "Cannot compute Jacobian while tape is being recorded."));
+          Assert(this->independent_variable_values.size() ==
+                   this->n_independent_variables(),
+                 ExcDimensionMismatch(this->independent_variable_values.size(),
+                                      this->n_independent_variables()));
+
+          this->taped_driver.jacobian(this->active_tape_index(),
+                                      this->n_dependent_variables(),
+                                      this->independent_variable_values,
+                                      jacobian);
+        }
+      else
+        {
+          Assert(ADNumberTraits<ad_type>::is_tapeless == true,
+                 ExcInternalError());
+          Assert(this->independent_variables.size() ==
+                   this->n_independent_variables(),
+                 ExcDimensionMismatch(this->independent_variables.size(),
+                                      this->n_independent_variables()));
+
+          this->tapeless_driver.jacobian(this->independent_variables,
+                                         this->dependent_variables,
+                                         jacobian);
+        }
+
+      for (unsigned int i = 0; i < this->n_dependent_variables(); i++)
+        for (unsigned int j = 0; j < this->n_independent_variables(); j++)
+          // Because we perform just a single differentiation
+          // operation with respect to the "column" variables,
+          // we only need to consider them for symmetry conditions.
+          if (this->is_symmetric_independent_variable(j) == true)
+            jacobian[i][j] *= 0.5;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    Tensor<0,
+           dim,
+           typename ADHelperVectorFunction<dim, ADNumberTypeCode, ScalarType>::
+             scalar_type>
+    ADHelperVectorFunction<dim, ADNumberTypeCode, ScalarType>::
+      extract_jacobian_component(
+        const FullMatrix<scalar_type> &   jacobian,
+        const FEValuesExtractors::Scalar &extractor_row,
+        const FEValuesExtractors::Scalar &extractor_col) const
+    {
+      // NOTE: It is necessary to make special provision for the case when the
+      // HessianType
+      //       is scalar. Unfortunately Tensor<0,dim> does not provide the
+      //       function unrolled_to_component_indices!
```

Comment:
funny indentation. On purpose?

Reply:
Also, there is no reason not to add that function to `Tensor<0>` if that would make your life easier!

Reply:
I've made a note in #7665 to do this.

Reply:
> funny indentation. On purpose?

Funny question :-D No, a certain tool (which shall remain anonymous) did something without me noticing...